src/session.c

```diff
@@ -611,21 +639,24 @@ static CK_RV slot_login(P11PROV_SLOT *slot, P11PROV_URI *uri,
     P11PROV_SESSION_POOL *pool = p11prov_slot_get_session_pool(slot);
     P11PROV_SESSION *session = NULL;
     CK_FLAGS flags = DEFLT_SESSION_FLAGS;
-    CK_STATE state = CKS_RO_PUBLIC_SESSION;
     int num_open_sessions = 0;
     CK_RV ret;
 
     /* try to get a login_session */
     ret = fetch_session(pool, flags, true, &session);
+    if (ret == CKR_USER_ALREADY_LOGGED_IN && _session == NULL) {
```

Comment:
This is now subtly behaving differently depending whether there's concurrent operations happening or not. The thread that gets the login session, will end up doing `session_check` other things, where as the contending thread losing access will not.

Perhaps this is intentional design as it will then make sure `session_check` is called periodically / often enough, but ensure race losing threads work too.

Reply:
So here is the rub of the problem.
The login operation can take arbitrarily long because you may be in a situation where a callback to the user is request to insert the PIN manually.

So the question is, what to do ?

Should we block all threads until this happens?
Perhaps we should ... I think I was worried about possible deadlocks in the past.

Maybe have a large timeout via backoff loop? So that eventually (say in 5 minutes) threads will unblock (and fail) ?

Reply:
So there's two cases:
- PIN is part of the URL or config, and the race is purely the time of what it takes to execute a login; usually less than few seconds
- PIN is not present, and user interaction is required; getting input can take arbitrary time

I think as along as the PIN provided case works, things should be good. The timeout should likely be a sane default (possibly configurable) on how much it takes time to provide PIN. I would probably think a minute should be good enough.

And I'd also make it so that the timeout count starts from the first operation initiated, and if concurrent threads join, they timeout when the first operation timeouts.